source/server/listener_manager_impl.cc

```diff
@@ -94,16 +96,95 @@ ProdListenerComponentFactory::createListenSocket(Network::Address::InstanceConst
   const int fd = server_.hotRestart().duplicateParentListenSocket(addr);
   if (fd != -1) {
     ENVOY_LOG(debug, "obtained socket for address {} from parent", addr);
-    return std::make_shared<Network::TcpListenSocket>(fd, address);
+    return std::make_shared<Network::TcpListenSocket>(fd, address, options);
   }
-  return std::make_shared<Network::TcpListenSocket>(address, bind_to_port);
+  return std::make_shared<Network::TcpListenSocket>(address, options, bind_to_port);
 }
 
 DrainManagerPtr
 ProdListenerComponentFactory::createDrainManager(envoy::api::v2::Listener::DrainType drain_type) {
   return DrainManagerPtr{new DrainManagerImpl(server_, drain_type)};
 }
 
+// Socket::Option implementation for API-defined listener socket options.
+// This same object can be extended to handle additional listener socket options.
+class ListenerSocketOption : public Network::Socket::Option, Logger::Loggable<Logger::Id::config> {
+public:
+  ListenerSocketOption(const envoy::api::v2::Listener& config)
+      : os_sys_calls_(Api::OsSysCallsSingleton::get()) {
+    if (config.has_transparent()) {
+      transparent_ = config.transparent().value();
+    }
+  }
+
+  // Network::Socket::Option
+  bool setOption(Network::Socket& socket, Network::Socket::SocketState state) const override {
+    // IP_TRANSPARENT has an effect on bind() (allowing bind() to non-local addresses), but
```

Comment:
Can you split this into a separate PR? I think it's orthogonal.

Reply:
Removed from this PR.